ksqldb-engine/src/main/java/io/confluent/ksql/function/udf/math/Greatest.java

```diff
@@ -0,0 +1,113 @@
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OF ANY KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package io.confluent.ksql.function.udf.math;
+
+import com.google.common.collect.Streams;
+import io.confluent.ksql.function.FunctionCategory;
+import io.confluent.ksql.function.udf.Udf;
+import io.confluent.ksql.function.udf.UdfDescription;
+import io.confluent.ksql.function.udf.UdfParameter;
+import io.confluent.ksql.function.udf.UdfSchemaProvider;
+import io.confluent.ksql.schema.ksql.SqlArgument;
+import io.confluent.ksql.schema.ksql.types.SqlType;
+import io.confluent.ksql.util.DecimalUtil;
+import io.confluent.ksql.util.KsqlConstants;
+import java.math.BigDecimal;
+import java.util.Arrays;
+import java.util.Comparator;
+import java.util.List;
+import java.util.Objects;
+import java.util.stream.Stream;
+
+@UdfDescription(
+    name = "greatest",
+    category = FunctionCategory.MATHEMATICAL,
+    description = "Returns the highest value among a variable number of consistently typed columns",
+    author = KsqlConstants.CONFLUENT_AUTHOR
+)
+public class Greatest {
+
+  @Udf
+  public Double greatest(@UdfParameter final Double val, @UdfParameter final Double... vals) {
+
+    final Stream<Double> toConcat = Stream.of(val);
+
+    return Streams.concat(toConcat, Arrays.stream(vals))
+        .filter(Objects::nonNull)
+        .max(Double::compareTo)
+        .orElse(null);
+  }
+
+  @Udf
+  public Integer greatest(@UdfParameter final Integer val, @UdfParameter final Integer... vals) {
+    final Stream<Integer> toConcat = Stream.of(val);
+    return Stream.concat(toConcat, Arrays.stream(vals))
+        .filter(Objects::nonNull)
+        .max(Integer::compareTo)
+        .orElse(null);
+  }
+
+  @Udf
+  public Long greatest(@UdfParameter final Long val, @UdfParameter final Long... vals) {
+    final Stream<Long> toConcat = Stream.of(val);
+    return Stream.concat(toConcat, Arrays.stream(vals))
+        .filter(Objects::nonNull)
+        .max(Long::compareTo)
+        .orElse(null);
+  }
+
+  @Udf
+  public String greatest(@UdfParameter final String val, @UdfParameter final String... vals) {
+
+    final Stream<String> toConcat = Stream.of(val);
+
+    return Streams.concat(toConcat, Arrays.stream(vals))
+        .filter(Objects::nonNull)
+        .max(String::compareTo)
+        .orElse(null);
+  }
+
+  @Udf(schemaProvider = "greatestDecimalProvider")
+  public BigDecimal greatest(@UdfParameter final BigDecimal val,
+      @UdfParameter final BigDecimal... vals) {
+
+    final Stream<BigDecimal> toConcat = Stream.of(val);
+
+    return Streams.concat(toConcat, Arrays.stream(vals))
+        .filter(Objects::nonNull)
+        .max(Comparator.naturalOrder())
+        .orElse(null);
+  }
+
+  @UdfSchemaProvider
+  public SqlType leastDecimalProvider(final List<SqlArgument> params) {
```

Comment:
This should be `greatestDecimalProvider` instead? I'm confused how this code is passing at all given that "greatestDecimalProvider" used in the annotation above doesn't exist.

Reply:
It was defaulting to the double function. I've added a test which checks for decimal places as I'm not immediately sure on if there's a way to test for types in QTTs? Regardless the next commit will have a test that just has too many decimal places for a double to accurately round to.